app/rdf/query-cube-preview.ts

```diff
@@ -0,0 +1,278 @@
+import uniqBy from "lodash/uniqBy";
+import { Quad } from "rdf-js";
+import ParsingClient from "sparql-http-client/ParsingClient";
+
+import {
+  BaseComponent,
+  BaseDimension,
+  DataCubePreview,
+  Dimension,
+  Measure,
+  TemporalDimension,
+} from "@/domain/data";
+import { truthy } from "@/domain/types";
+import { resolveDimensionType, resolveMeasureType } from "@/graphql/resolvers";
+
+import * as ns from "./namespace";
+import {
+  getDataKind,
+  getIsNumerical,
+  getScaleType,
+  getTimeFormat,
+  getTimeUnit,
+  parseNumericalTerm,
+  parseResolution,
+} from "./parse";
+import { buildLocalizedSubQuery } from "./query-utils";
+
+// ?s ?p ?o
+type ShortQuad = Omit<Quad, "subject" | "predicate" | "object"> & {
+  s: Quad["subject"];
+  p: Quad["predicate"];
+  o: Quad["object"];
+};
+
+export const getCubePreview = async (
+  iri: string,
+  options: {
+    locale: string;
+    latest: Boolean;
+    sparqlClient: ParsingClient;
+  }
+): Promise<DataCubePreview> => {
+  const { sparqlClient, locale } = options;
+  const qs: ShortQuad[] = await sparqlClient.query
+    .construct(
+      `PREFIX cube: <https://cube.link/>
+PREFIX meta: <https://cube.link/meta/>
+PREFIX qudt: <http://qudt.org/schema/qudt/>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX schema: <http://schema.org/>
+PREFIX sh: <http://www.w3.org/ns/shacl#>
+PREFIX time: <http://www.w3.org/2006/time#>
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+    
+CONSTRUCT {
+  ?cube schema:version ?cubeVersion .
+  ?cube cube:observationSet ?observationSet .
+  ?observationSet cube:observation ?observation .
+
+  ?dimension sh:path ?dimensionIri .
+  ?dimension sh:datatype ?dimensionDataType .
+  ?dimension rdf:type ?dimensionType .
+  ?dimension qudt:scaleType ?dimensionScaleType .
+  ?dimension qudt:unit ?dimensionUnit .
+  ?dimensionUnit schema:name ?dimensionUnitLabel .
+  ?dimensionUnit qudt:CurrencyUnit ?dimensionUnitIsCurrency .
+  ?dimensionUnit qudt:currencyExponent ?dimensionUnitCurrencyExponent .
+  ?dimension sh:order ?dimensionOrder .
+  ?dimension meta:dataKind ?dimensionDataKind .
+  ?dimensionDataKind rdf:type ?dimensionDataKindType .
+  ?dimensionDataKind time:unitType ?dimensionTimeUnitType .
+  ?dimension schema:name ?dimensionLabel .
+  ?dimension schema:description ?dimensionDescription .
+
+  ?observation ?observationPredicate ?observationValue .
+  ?observation ?observationPredicate ?observationValueLabel .
+} WHERE {
+  VALUES ?cube { <${iri}> }
+  FILTER(EXISTS { ?cube a cube:Cube . }) {}
+  UNION {
+    ?cube cube:observationConstraint/sh:property ?dimension .
+    ?dimension sh:path ?dimensionIri .
+    OPTIONAL { ?dimension sh:datatype ?dimensionDataType . }
```

Comment:
in case datatype is effectively needed, it can be in two places - the other one being
`    OPTIONAL { ?dimension sh:or/rdf:rest*/rdf:first/sh:datatype ?dimensionDataType . }`

Reply:
@Rdataflow I tried to include the additional line to SPARQL query and it seems to not affect the performance in noticeable way 👍 I think we might stay with it for the moment if that's fine with you :)

Reply:
In fact, when trying with a different, bigger cube, the impact feel more significant. I'll take a look tomorrow to retrieve data type directly from observations 👀